src/java/io/bazel/rulesscala/test_discovery/FilteredRunnerBuilder.scala

```diff
@@ -21,29 +28,40 @@ class FilteredRunnerBuilder(builder: RunnerBuilder, filteringRunnerBuilder: Filt
   }
 }
 
-object FilteredRunnerBuilder {
-  type FilteringRunnerBuilder = PartialFunction[(Runner, Class[_], Pattern), Runner]
+private[rulesscala] class FilteredTestClass(testClass: Class[_], pattern: Pattern) extends TestClass(testClass) {
+  override def getAnnotatedMethods(aClass: Class[_ <: Annotation]): util.List[FrameworkMethod] = {
+    val methods = super.getAnnotatedMethods(aClass)
+    if (aClass == classOf[Test]) methods.filter(method => methodMatchesPattern(method, pattern))
+    else methods
+  }
+
+  private def methodMatchesPattern(method: FrameworkMethod, pattern: Pattern): Boolean = {
+    val testCase = method.getDeclaringClass.getName + "#" + method.getName
+    pattern.matcher(testCase).find
+  }
 }
 
 object JUnitFilteringRunnerBuilder {
   val f: FilteringRunnerBuilder = {
-    case (_: BlockJUnit4ClassRunner, testClass: Class[_], pattern: Pattern) =>
-      new FilteredJUnitClassRunner(testClass, pattern)
+    case (runner: BlockJUnit4ClassRunner, testClass: Class[_], pattern: Pattern) =>
+      replaceRunnerTestClass(runner, testClass, pattern)
   }
-}
 
-class FilteredJUnitClassRunner(testClass: Class[_], pattern: Pattern)
-  extends BlockJUnit4ClassRunner(testClass) {
-  override def getChildren: java.util.List[FrameworkMethod] =
-    super
-      .getChildren
-      .asScala
-      .filter(method => methodMatchesPattern(method, pattern))
-      .asJava
-
-  private def methodMatchesPattern(method: FrameworkMethod, pattern: Pattern): Boolean = {
-    val testCase = method.getDeclaringClass.getName + "#" + method.getName
-    pattern.matcher(testCase).find
+  private def replaceRunnerTestClass(runner: BlockJUnit4ClassRunner, testClass: Class[_], pattern: Pattern) = {
+    allFieldsOf(runner.getClass)
+      .find(f => f.getName == "testClass" || f.getName == "fTestClass")
```

Comment:
Can we comment why runner has these fields and why we are testing two different names? Is that to deal with different versions?

Reply:
@hmemcpy can you please add the comment @johnynek asked for?

Reply:
Oops, missed it. Yes, it has to do with JUnit version, the field was renamed in JUnit 4.12.

Reply:
Done, I extracted those values into named fields, instead of a comment. @johnynek please review :)